aggregate/billing_functions/gcp.py

```diff
@@ -0,0 +1,164 @@
+
+Tasks:
+
+- Needs to convert {billing_project_id} into DATSET
+- Only want to transfer data from the projects in the server-config
+- Can't duplicate rows, so:
+    - just grab only settled data within START + END of previous time period
+- Service ID should be faithfully handed over
+- Should search and update for [START_PERIOD, END_PERIOD)
+"""
+
+import json
+import hashlib
+
+from datetime import datetime
+from typing import Dict
+
+from pandas import DataFrame
+import google.cloud.bigquery as bq
+
+try:
+    from . import utils
+except ImportError:
+    import utils
+
+
+logger = utils.logger
+
+
+##########################
+#    INPUT PROCESSORS    #
+##########################
+
+
+def from_request(request):
+    """
+    From request object, get start and end time if present
+    """
+    start, end = utils.get_start_and_end_from_request(request)
+    main(start, end)
+
+
+def from_pubsub(data=None, _=None):
+    """
+    From pubsub message, get start and end time if present
+    """
+    start, end = utils.get_start_and_end_from_data(data)
+    main(start, end)
+
+
+#################
+#    MIGRATE    #
+#################
+
+
+def migrate_billing_data(start, end, dataset_to_gcp_map) -> int:
+    """
+    Get the billing date in the time period
+    Filter out any rows that aren't in the allowed project ids
+    """
+
+    def get_topic(row):
+        return utils.billing_row_to_topic(row, dataset_to_gcp_map)
+
+    migrate_rows = get_billing_data(start, end)
+
+    if len(migrate_rows) == 0:
+        logger.info('No rows to migrate')
+        return 0
+
+    # Add id and dataset to the row
+    migrate_rows = migrate_rows.drop(columns=['billing_account_id'])
+    migrate_rows.insert(0, 'topic', migrate_rows.apply(get_topic, axis=1))
+    migrate_rows.insert(0, 'id', migrate_rows.apply(billing_row_to_key, axis=1))
+
+    result = utils.insert_dataframe_rows_in_table(
+        utils.GCP_AGGREGATE_DEST_TABLE, migrate_rows
+    )
+
+    return result
+
+
+#################
+#    HELPERS    #
+#################
+
+
+def get_billing_data(start: datetime, end: datetime) -> DataFrame:
+    """
+    Retrieve the billing data from start to end date inclusive
+    Return results as a dataframe
+    """
+
+    _query = f"""
+        SELECT * FROM `{utils.GCP_BILLING_BQ_TABLE}`
+        WHERE export_time >= @start
+            AND export_time <= @end
+    """
+    job_config = bq.QueryJobConfig(
+        query_parameters=[
+            bq.ScalarQueryParameter('start', 'STRING', str(start)),
+            bq.ScalarQueryParameter('end', 'STRING', str(end)),
+        ]
+    )
+
+    migrate_rows = (
+        utils.get_bigquery_client()
+        .query(_query, job_config=job_config)
+        .result()
+        .to_dataframe()
+    )
+
+    return migrate_rows
+
+
+def billing_row_to_key(row) -> str:
+    """Convert a billing row to a hash which will be the row key"""
+    data = tuple(row)
+    identifier = hashlib.md5()
```

Comment:
Why do we need a hash, compared to e.g. just an auto-incrementing index?

Reply:
The hash as a key means we don't migrate the same data twice as these functions (so these functions are all idempotent).